File: backend/app/services/money.py
"""Single source of truth for the points ↔ naira conversion.

Every place in the backend that credits or debits a user's
`points_balance` goes through one of these helpers. The conversion
rate itself is owned by `settings.points_per_naira` (env var
`POINTS_PER_NAIRA`) so that revaluing the in-app currency is a
one-line change in `.env` instead of a sweep through every credit
site.

Convention (units):
  - kobo     = smallest NGN unit (100 kobo = ₦1)
  - naira    = NGN whole unit (1 naira = 100 kobo)
  - points   = in-app wallet unit (POINTS_PER_NAIRA points = ₦1)

Default rate: 10 points per ₦1 (1 point = ₦0.10). To revalue,
override POINTS_PER_NAIRA in `backend/.env`.

The mobile client reads EXPO_PUBLIC_POINTS_PER_NAIRA at build time
with the same value, so both sides always agree.
"""

from app.config import settings


def kobo_to_naira(kobo: int) -> float:
    """Convert kobo (smallest NGN unit) to whole naira.

    Example: kobo_to_naira(10000) == 100.0  (₦100.00)
    """
    return kobo / 100


def kobo_to_points(kobo: int) -> int:
    """Convert kobo to wallet points using the configured rate.

    This is the credit side: when a worker is owed ₦100 (10000 kobo)
    and POINTS_PER_NAIRA=10, the worker gets 1000 points.
    Example: kobo_to_points(10000) == 1000
    """
    return int(round((kobo / 100) * settings.points_per_naira))


def naira_to_points(naira: float) -> int:
    """Convert whole naira to wallet points.

    Example: naira_to_points(100) == 1000   (when POINTS_PER_NAIRA=10)
    """
    return int(round(naira * settings.points_per_naira))


def points_to_naira(points: int) -> float:
    """Convert wallet points back to whole naira.

    Used for display-only paths (worker UI already reads the same
    conversion via EXPO_PUBLIC_POINTS_PER_NAIRA).
    Example: points_to_naira(1000) == 100.0  (when POINTS_PER_NAIRA=10)
    """
    return points / settings.points_per_naira


def points_to_kobo(points: int) -> int:
    """Convert wallet points back to kobo.

    Inverse of `kobo_to_points`. Used by payouts/withdrawals when
    the user picks "cash out X points".
    Example: points_to_kobo(1000) == 10000   (when POINTS_PER_NAIRA=10)
    """
    return int(round((points / settings.points_per_naira) * 100))

File: client/src/shared/lib/money.ts
/**
 * Money formatting + points ↔ naira conversion for PagePay.
 *
 * Three units in play throughout the app:
 *   - kobo     : smallest NGN unit (100 kobo = ₦1)
 *   - naira    : NGN whole unit (1 naira = 100 kobo)
 *   - points   : in-app wallet unit (POINTS_PER_NAIRA points = ₦1)
 *
 * The conversion rate `POINTS_PER_NAIRA` is configured once per
 * environment (`.env` files) so a future revaluation is a one-line
 * change:
 *   backend/.env     → POINTS_PER_NAIRA=10
 *   client/.env      → EXPO_PUBLIC_POINTS_PER_NAIRA=10
 *
 * Both sides must agree, otherwise the worker will see one number in
 * the wallet header and a different number on the task card. The
 * helpers below guarantee both display paths use the same number.
 */

/** Read the points-per-naira rate from the Expo public env. Default 10. */
const POINTS_PER_NAIRA = parseInt(
  // EXPO_PUBLIC_* is inlined at build time; falls back to 10 if the
  // env var is missing (e.g. local dev before .env was copied).
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (typeof process !== 'undefined' ? process.env?.EXPO_PUBLIC_POINTS_PER_NAIRA : null) ?? '10',
  10,
);

/** NGN value of one wallet point. Derived from the env rate, not hardcoded. */
export const NGN_PER_POINT = 1 / POINTS_PER_NAIRA;

/** Points value of one whole NGN. Derived from the env rate, not hardcoded. */
export const POINTS_PER_NAIRA_VALUE = POINTS_PER_NAIRA;

/**
 * Format a kobo amount as a Naira string with two decimal places and
 * thousands separators. Examples (with POINTS_PER_NAIRA=10):
 *   formatKobo(0)        → "₦0.00"
 *   formatKobo(1500)     → "₦15.00"
 *   formatKobo(123456)   → "₦1,234.56"
 *   formatKobo(1_000_000) → "₦10,000.00"
 *
 * 1 kobo = ₦0.01 (a fixed NGN convention, NOT derived from
 * POINTS_PER_NAIRA — kobo is a currency sub-unit, independent of the
 * in-app points rate).
 */
export function formatKobo(kobo: number): string {
  const ngn = kobo / 100;
  // toFixed(2) is locale-independent and always emits "123456.78" — we
  // split on "." and add the thousands separator manually so the output
  // is stable across Hermes versions.
  const [whole, frac = ''] = ngn.toFixed(2).split('.');
  const withSep = (whole ?? '0').replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  return `₦${withSep}.${frac.padEnd(2, '0')}`;
}

/**
 * Format a points balance for display in the wallet header / history.
 * Plain thousands-separated integer. Examples:
 *   formatPoints(0)        → "0"
 *   formatPoints(1234)     → "1,234"
 *   formatPoints(1_000_000) → "1,000,000"
 */
export function formatPoints(points: number): string {
  return points.toLocaleString();
}

/**
 * Compact points formatter for tight UI like the home header chip.
 * Rounds to one decimal and appends k / m. Examples:
 *   formatPointsCompact(0)       → "0"
 *   formatPointsCompact(999)     → "999"
 *   formatPointsCompact(1000)    → "1k"
 *   formatPointsCompact(1100)    → "1.1k"
 *   formatPointsCompact(10_000)  → "10k"
 *   formatPointsCompact(1_200_000) → "1.2m"
 */
export function formatPointsCompact(points: number): string {
  const abs = Math.abs(points);
  if (abs >= 1_000_000) {
    const val = points / 1_000_000;
    return `${val.toFixed(2)}m`;
  }
  if (abs >= 1000) {
    const val = points / 1000;
    return `${val.toFixed(2)}k`;
  }
  return String(points);
}

/**
 * Convert a kobo-denominated reward (e.g. `task.reward_amount` from
 * the API, or `submission.reward_paid`) into the wallet points the
 * worker will be credited. With POINTS_PER_NAIRA=10:
 *   koboToPoints(10000)   → 1000    (₦100 reward = 1,000 pts)
 *   koboToPoints(50000)   → 5000    (₦500 reward = 5,000 pts)
 */
export function koboToPoints(kobo: number): number {
  return Math.round((kobo / 100) * POINTS_PER_NAIRA);
}

/**
 * Convert a kobo-denominated reward into the naira string the worker
 * sees on the task card. Thin wrapper around `formatKobo` kept here
 * so call sites read like the rest of the conversions.
 */
export function koboToNairaString(kobo: number): string {
  return formatKobo(kobo);
}

/**
 * Convert wallet points back to the naira string. With
 * POINTS_PER_NAIRA=10:
 *   pointsToNairaString(1000) → "₦100.00"
 *   pointsToNairaString(5000) → "₦500.00"
 */
export function pointsToNairaString(points: number): string {
  return formatKobo(points * (100 / POINTS_PER_NAIRA));
}

/**
 * Withdrawal fee tier table. Duplicated from
 * `backend/app/config.py:Settings.withdrawal_fee_tiers` (the default
 * schedule) so the modal can show a live fee preview without a config
 * roundtrip.
 *
 *   ≤ ₦5,000          → ₦15  (Paystack ₦10, profit ₦5)
 *   ₦5,001 – ₦50,000  → ₦35  (Paystack ₦25, profit ₦10)
 *   > ₦50,000         → ₦70  (Paystack ₦50, profit ₦20)
 *
 * The server re-computes the fee on submit and is the source of truth;
 * this is only for the live UI preview.
 */
export const WITHDRAWAL_FEE_TIERS_KOBO: readonly (readonly [number | null, number])[] = [
  [500_000, 1_500],
  [5_000_000, 3_500],
  [null, 7_000],
];

/** Return the user-paid flat fee (in kobo) for a withdrawal of `amountKobo`. */
export function previewWithdrawalFeeKobo(amountKobo: number): number {
  for (const [maxKobo, feeKobo] of WITHDRAWAL_FEE_TIERS_KOBO) {
    if (maxKobo === null || amountKobo <= maxKobo) return feeKobo;
  }
  return WITHDRAWAL_FEE_TIERS_KOBO[WITHDRAWAL_FEE_TIERS_KOBO.length - 1]?.[1] ?? 0;
}